File: src/ffxiv/fisher/model/Area.java
package ffxiv.fisher.model;

public enum Area {
	
	// La Noscea
	LIMSA_LOMINSA_UPPER_DECKS(0, Region.LA_NOSCEA),
	LIMSA_LOMINSA_LOWER_DECKS(1, Region.LA_NOSCEA),
	MIDDLE_LA_NOSCEA(2, Region.LA_NOSCEA),
	LOWER_LA_NOSCEA(3, Region.LA_NOSCEA),
	EASTERN_LA_NOSCEA(4, Region.LA_NOSCEA),
	WESTERN_LA_NOSCEA(5, Region.LA_NOSCEA),
	UPPER_LA_NOSCEA(6, Region.LA_NOSCEA),
	OUTER_LA_NOSCEA(7, Region.LA_NOSCEA),
	MIST(8, Region.LA_NOSCEA),
	
	// The Black Shroud
	NEW_GRIDANIA(9, Region.THE_BLACK_SHROUD),
	OLD_GRIDANIA(10, Region.THE_BLACK_SHROUD),
	CENTRAL_SHROUD(11, Region.THE_BLACK_SHROUD),
	EAST_SHROUD(12, Region.THE_BLACK_SHROUD),
	SOUTH_SHROUD(13, Region.THE_BLACK_SHROUD),
	NORTH_SHROUD(14, Region.THE_BLACK_SHROUD),
	LAVENDER_BEDS(15, Region.THE_BLACK_SHROUD),
	
	// Thanalan
	WESTERN_THANALAN(16, Region.THANALAN),
	CENTRAL_THANALAN(17, Region.THANALAN),
	EASTERN_THANALAN(18, Region.THANALAN),
	SOUTHERN_THANALAN(19, Region.THANALAN),
	NORTHERN_THANALAN(20, Region.THANALAN),
	THE_GOBLET(21, Region.THANALAN),
	
	// Coerthas
	COERTHAS_CENTRAL_HIGHLANDS(22, Region.COERTHAS),
	
	// Mor Dhona
	MOR_DHONA(23, Region.MOR_DHONA);
	
	
	
	private final int clientIdentifier;
	private final Region region;

	private Area(int clientIdentifier, Region region) {
		this.clientIdentifier = clientIdentifier;
		this.region = region;
	}

	public Region getRegion() {
		return region;
	}
	
	public int getClientIdentifier() {
		return clientIdentifier;
	}
}

File: js_src/model/area.js

goog.provide('ff.model.Area');

goog.require('ff.model.Region');
goog.require('goog.object');
goog.require('goog.string');
goog.provide('ff.model.AreaEnum');



/**
 * @param {number} clientIdentifier
 * @param {!ff.model.Region} region
 * @param {string} name
 * @constructor
 */
ff.model.Area = function(clientIdentifier, region, name) {

  /** @private {number} */
  this.clientIdentifier_ = clientIdentifier;

  /** @private {!ff.model.Region} */
  this.region_ = region;

  /** @private {string} */
  this.name_ = name;
};


/** @return {number} */
ff.model.Area.prototype.getClientIdentifier = function() {
  return this.clientIdentifier_;
};


/** @return {!ff.model.Region} */
ff.model.Area.prototype.getRegion = function() {
  return this.region_;
};


/** @return {string} */
ff.model.Area.prototype.getName = function() {
  return this.name_;
};


/**
 * Figures out the enum given the model.
 * @param {!ff.model.Area} area
 * @return {string}
 */
ff.model.Area.getEnum = function(area) {
  var key = goog.object.findKey(
      ff.model.AreaEnum,
      function(value, key, object) {
        return goog.string.caseInsensitiveCompare(
            value.getName(), area.getName()) == 0;
      });
  if (!key) {
    throw Error('Unable to find AreaEnum for ' + area.getName());
  }
  return key;
};


/**
 * Figures out the enum given the identifier.
 * @param {number} identifier
 * @return {string}
 */
ff.model.Area.getEnumFromIdentifier = function(identifier) {
  var key = goog.object.findKey(
      ff.model.AreaEnum,
      function(value, key, object) {
        return value.getClientIdentifier() == identifier;
      });
  if (!key) {
    throw Error('Unable to find AreaEnum for identifier ' + identifier);
  }
  return key;
};


/**
 * @enum {!ff.model.Area}
 */
ff.model.AreaEnum = {
  // La Noscea
  'LIMSA_LOMINSA_UPPER_DECKS': new ff.model.Area(
      0, ff.model.Region.LA_NOSCEA, 'Limsa Lominsa Upper Decks'),
  'LIMSA_LOMINSA_LOWER_DECKS': new ff.model.Area(
      1, ff.model.Region.LA_NOSCEA, 'Limsa Lominsa Lower Decks'),
  'MIDDLE_LA_NOSCEA': new ff.model.Area(
      2, ff.model.Region.LA_NOSCEA, 'Middle La Noscea'),
  'LOWER_LA_NOSCEA': new ff.model.Area(
      3, ff.model.Region.LA_NOSCEA, 'Lower La Noscea'),
  'EASTERN_LA_NOSCEA': new ff.model.Area(
      4, ff.model.Region.LA_NOSCEA, 'Eastern La Noscea'),
  'WESTERN_LA_NOSCEA': new ff.model.Area(
      5, ff.model.Region.LA_NOSCEA, 'Western La Noscea'),
  'UPPER_LA_NOSCEA': new ff.model.Area(
      6, ff.model.Region.LA_NOSCEA, 'Upper La Noscea'),
  'OUTER_LA_NOSCEA': new ff.model.Area(
      7, ff.model.Region.LA_NOSCEA, 'Outer La Noscea'),
  'MIST': new ff.model.Area(
      8, ff.model.Region.LA_NOSCEA, 'Mist'),

  // The Black Shroud
  'NEW_GRIDANIA': new ff.model.Area(
      9, ff.model.Region.THE_BLACK_SHROUD, 'New Gridania'),
  'OLD_GRIDANIA': new ff.model.Area(
      10, ff.model.Region.THE_BLACK_SHROUD, 'Old Gridania'),
  'CENTRAL_SHROUD': new ff.model.Area(
      11, ff.model.Region.THE_BLACK_SHROUD, 'Central Shroud'),
  'EAST_SHROUD': new ff.model.Area(
      12, ff.model.Region.THE_BLACK_SHROUD, 'East Shroud'),
  'SOUTH_SHROUD': new ff.model.Area(
      13, ff.model.Region.THE_BLACK_SHROUD, 'South Shroud'),
  'NORTH_SHROUD': new ff.model.Area(
      14, ff.model.Region.THE_BLACK_SHROUD, 'North Shroud'),
  'LAVENDER_BEDS': new ff.model.Area(
      15, ff.model.Region.THE_BLACK_SHROUD, 'Lavender Beds'),

  // Thanalan
  'WESTERN_THANALAN': new ff.model.Area(
      16, ff.model.Region.THANALAN, 'Western Thanalan'),
  'CENTRAL_THANALAN': new ff.model.Area(
      17, ff.model.Region.THANALAN, 'Central Thanalan'),
  'EASTERN_THANALAN': new ff.model.Area(
      18, ff.model.Region.THANALAN, 'Eastern Thanalan'),
  'SOUTHERN_THANALAN': new ff.model.Area(
      19, ff.model.Region.THANALAN, 'Southern Thanalan'),
  'NORTHERN_THANALAN': new ff.model.Area(
      20, ff.model.Region.THANALAN, 'Northern Thanalan'),
  'THE_GOBLET': new ff.model.Area(
      21, ff.model.Region.THANALAN, 'The Goblet'),

  // Coerthas
  'COERTHAS_CENTRAL_HIGHLANDS': new ff.model.Area(
      22, ff.model.Region.COERTHAS, 'Coerthas Central Highlands'),

  // Mor Dhona
  'MOR_DHONA': new ff.model.Area(
      23, ff.model.Region.MOR_DHONA, 'Mor Dhona')
};
